File: 2025/day12/php/solution.php
#!/usr/bin/env php
<?php
/**
 * Day 12: Christmas Tree Farm - Polyomino Packing
 *
 * The solution checks if presents (polyominoes) can fit into rectangular regions.
 * For this problem, the constraint is simply: total cells needed <= available cells.
 */

declare(strict_types=1);

function parseInput(string $text): array {
    $sections = array_filter(array_map(trim(...), explode("\n\n", trim($text))));

    $shapes = [];
    $regions = [];

    foreach ($sections as $section) {
        $lines = array_filter(array_map(trim(...), explode("\n", $section)));

        if (empty($lines)) {
            continue;
        }

        $firstLine = $lines[0];

        if (str_contains($firstLine, ':') && !str_contains($firstLine, 'x')) {
            // Shape definition
            $idx = (int) rtrim($firstLine, ':');
            $shapeLines = array_slice($lines, 1);

            // Count total '#' cells in the shape using array functions
            $cellCount = array_sum(
                array_map(fn($line) => substr_count($line, '#'), $shapeLines)
            );

            $shapes[$idx] = $cellCount;
        } else {
            // Region definitions
            foreach ($lines as $line) {
                if (str_contains($line, 'x')) {
                    [$dims, $countsStr] = explode(':', $line, 2);
                    [$w, $h] = array_map('intval', explode('x', trim($dims)));
                    $counts = array_map('intval', preg_split('/\s+/', trim($countsStr)));
                    $regions[] = [$w, $h, $counts];
                }
            }
        }
    }

    return [$shapes, $regions];
}

function canFitRegion(int $width, int $height, array $counts, array $shapeSizes): bool {
    // Calculate total cells needed using array_sum with array_map
    $totalCellsNeeded = array_sum(
        array_map(
            fn($i) => $counts[$i] * $shapeSizes[$i],
            array_keys($counts)
        )
    );

    $available = $width * $height;
    return $totalCellsNeeded <= $available;
}

function part1(array $shapes, array $regions): int {
    // Count regions that can fit using array_filter
    return count(
        array_filter(
            $regions,
            fn($region) => canFitRegion($region[0], $region[1], $region[2], $shapes)
        )
    );
}

function part2(): int {
    // Part 2 is just a button click to finish - no computation needed
    return 0;
}

// Main execution
$text = file_get_contents('../input.txt');
[$shapes, $regions] = parseInput($text);

echo "Part 1: " . part1($shapes, $regions) . "\n";
echo "Part 2: " . part2() . "\n";

File: 2025/day12/perl/solution.pl
#!/usr/bin/env perl
use strict;
use warnings;
use v5.10;
use List::Util qw(sum);

# Day 12: Christmas Tree Farm - Polyomino Packing
#
# The solution checks if presents (polyominoes) can fit into rectangular regions.
# For this problem, the constraint is simply: total cells needed <= available cells.

sub parse_input {
    my ($text) = @_;
    my %shapes;
    my @regions;

    # Split by double newlines to get sections
    my @sections = split /\n\n/, $text;

    foreach my $section (@sections) {
        my @lines = split /\n/, $section;
        next unless @lines;

        if ($lines[0] =~ /^(\d+):$/ && $lines[0] !~ /x/) {
            # Shape definition - count '#' characters using map and scalar context
            my $idx = $1;
            my $cell_count = sum map { scalar(() = /\#/g) } @lines[1..$#lines];
            $shapes{$idx} = $cell_count;
        } else {
            # Region definitions - use map to extract from lines
            push @regions, map {
                /(\d+)x(\d+):\s*(.+)/ ? {
                    width => $1,
                    height => $2,
                    counts => [split /\s+/, $3]
                } : ()
            } @lines;
        }
    }

    return (\%shapes, \@regions);
}

sub can_fit_region {
    my ($width, $height, $counts, $shape_sizes) = @_;

    # Use map to calculate cells needed for each shape, then sum
    my $total_cells_needed = sum map {
        $counts->[$_] * $shape_sizes->{$_}
    } 0..$#{$counts};

    my $available = $width * $height;
    return $total_cells_needed <= $available;
}

sub part1 {
    my ($shapes, $regions) = @_;

    # Use grep to count regions that fit
    return scalar grep {
        can_fit_region($_->{width}, $_->{height}, $_->{counts}, $shapes)
    } @$regions;
}

sub part2 {
    # Part 2 is just a button click to finish - no computation needed
    return 0;
}

# Main execution
open my $fh, '<', '../input.txt' or die "Cannot open input.txt: $!";
my $text = do { local $/; <$fh> };
close $fh;

my ($shapes, $regions) = parse_input($text);

say "Part 1: ", part1($shapes, $regions);
say "Part 2: ", part2();
